## Setup

In [ ]:
import pandas as pd

In [ ]:
ufos_path = 'YOUR PATH HERE'

In [ ]:
sightings = pd.read_csv(ufos_path)
sightings.info()

<ipython-input-17-01ca1588f900>:1: DtypeWarning: Columns (5,9) have mixed types. Specify dtype option on import or set low_memory=False.
  sightings = pd.read_csv(ufos_path)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              80332 non-null  object 
 1   city                  80332 non-null  object 
 2   state                 74535 non-null  object 
 3   country               70662 non-null  object 
 4   shape                 78400 non-null  object 
 5   duration (seconds)    80332 non-null  object 
 6   duration (hours/min)  80332 non-null  object 
 7   comments              80317 non-null  object 
 8   date posted           80332 non-null  object 
 9   latitude              80332 non-null  object 
 10  longitude             80332 non-null  float64
dtypes: float64(1), object(10)
memory usage: 6.7+ MB


## Remove all rows from the UFO sightings dataset with a `Null` value in the "state" column:

In [ ]:
no_null_states = sightings.dropna(subset='state')
no_null_states.info()

<class 'pandas.core.frame.DataFrame'>
Index: 74535 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              74535 non-null  object 
 1   city                  74535 non-null  object 
 2   state                 74535 non-null  object 
 3   country               68121 non-null  object 
 4   shape                 72742 non-null  object 
 5   duration (seconds)    74535 non-null  object 
 6   duration (hours/min)  74535 non-null  object 
 7   comments              74522 non-null  object 
 8   date posted           74535 non-null  object 
 9   latitude              74535 non-null  object 
 10  longitude             74535 non-null  float64
dtypes: float64(1), object(10)
memory usage: 6.8+ MB


## How many rows were removed from the DataFrame?

In [ ]:
len(sightings) - len(no_null_states)

5797

**5,797**

## Remove all rows from the UFO sightings dataset with a `Null` value in either the "state" OR "country" columns:

In [ ]:
sightings.dropna(subset=['state','country'], inplace=True)
sightings.info()

<class 'pandas.core.frame.DataFrame'>
Index: 68121 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              68121 non-null  object 
 1   city                  68121 non-null  object 
 2   state                 68121 non-null  object 
 3   country               68121 non-null  object 
 4   shape                 66524 non-null  object 
 5   duration (seconds)    68121 non-null  object 
 6   duration (hours/min)  68121 non-null  object 
 7   comments              68108 non-null  object 
 8   date posted           68121 non-null  object 
 9   latitude              68121 non-null  object 
 10  longitude             68121 non-null  float64
dtypes: float64(1), object(10)
memory usage: 6.2+ MB


## How many rows in the original DataFrame had `Null` values in the "country" column but not the "state" column?

In [ ]:
len(no_null_states) - len(sightings)

6414

**6,414**

## Preview the original DataFrame:


In [ ]:
sightings.head()

datetime        city state country     shape duration (seconds)  \
0  10/10/1949 20:30  san marcos    tx      us  cylinder               2700   
3  10/10/1956 21:00        edna    tx      us    circle                 20   
4  10/10/1960 20:00     kaneohe    hi      us     light                900   
5  10/10/1961 19:00     bristol    tn      us    sphere                300   
7  10/10/1965 23:45     norwalk    ct      us      disk               1200   

  duration (hours/min)                                           comments  \
0           45 minutes  This event took place in early fall around 194...   
3             1/2 hour  My older brother and twin sister were leaving ...   
4           15 minutes  AS a Marine 1st Lt. flying an FJ4B fighter/att...   
5            5 minutes  My father is now 89 my brother 52 the girl wit...   
7           20 minutes  A bright orange color changing to reddish colo...   

  date posted    latitude   longitude  
0   4/27/2004  29.8830556  -97.941111  
3   1/17/2004  28.9783333  -96.645833  
4   1/22/2004  21.4180556 -157.803611  
5   4/27/2007  36.5950000  -82.188889  
7   10/2/1999  41.1175000  -73.408333

## Reset the index of the DataFrame and preview it again:

In [ ]:
sightings.reset_index(inplace=True)
sightings.head()

index          datetime        city state country     shape  \
0      0  10/10/1949 20:30  san marcos    tx      us  cylinder   
1      3  10/10/1956 21:00        edna    tx      us    circle   
2      4  10/10/1960 20:00     kaneohe    hi      us     light   
3      5  10/10/1961 19:00     bristol    tn      us    sphere   
4      7  10/10/1965 23:45     norwalk    ct      us      disk   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1                 20             1/2 hour   
2                900           15 minutes   
3                300            5 minutes   
4               1200           20 minutes   

                                            comments date posted    latitude  \
0  This event took place in early fall around 194...   4/27/2004  29.8830556   
1  My older brother and twin sister were leaving ...   1/17/2004  28.9783333   
2  AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004  21.4180556   
3  My father is now 89 my brother 52 the girl wit...   4/27/2007  36.5950000   
4  A bright orange color changing to reddish colo...   10/2/1999  41.1175000   

    longitude  
0  -97.941111  
1  -96.645833  
2 -157.803611  
3  -82.188889  
4  -73.408333